mopidy_spotify/playlists.py

```diff
@@ -65,16 +144,114 @@ def refresh(self):
 
     def delete(self, uri):
-        pass  # TODO
+        logger.info(f"Deleting playlist {uri}")
+        ok = web.delete_playlist(self._backend._web_client, uri)
+        self._get_flattened_playlist_refs()
+        return ok
+
+    @staticmethod
+    def _len_replace(playlist, n=_chunk_size):
+        return math.ceil(len(playlist.tracks) / n)
+
+    @staticmethod
+    def _is_spotify_track(track_uri):
+        return track_uri.startswith("spotify:track:")
 
     def save(self, playlist):
-        pass  # TODO
+        saved_playlist = self._get_playlist(playlist.uri, with_owner=True)
+        if not saved_playlist:
+            return
+
+        saved_playlist, owner = saved_playlist
+        # We limit playlist editing to the user's own playlists, since mopidy
+        # mangles the names of other people's playlists.
+        if owner and owner != self._backend._web_client.user_id:
+            logger.error(
+                f"Refusing to modify someone else's playlist ({playlist.uri})"
+            )
+            return
+
+        # We cannot add or (easily) remove spotify:local: tracks, so refuse
+        # editing if the current playlist contains such tracks.
+        if any(
+            (t.uri.startswith("spotify:local:") for t in saved_playlist.tracks)
+        ):
+            logger.error(
+                "Cannot modify playlist containing 'Spotify Local Files'."
+            )
+            for t in saved_playlist.tracks:
+                if t.uri.startswith("spotify:local:"):
+                    logger.error(f"{t.name} ({t.uri})")
+            return
+
+        new_tracks = [track.uri for track in playlist.tracks]
+        cur_tracks = [track.uri for track in saved_playlist.tracks]
+
+        if any((not self._is_spotify_track(t) for t in new_tracks)):
+            new_tracks = list(filter(self._is_spotify_track, new_tracks))
+            logger.warning(
+                "Cannot add non-spotify tracks to spotify playlist; skipping those."
+            )
+
+        operations = utils.diff(cur_tracks, new_tracks, self._chunk_size)
+
+        # calculate number of operations required for each strategy:
+        ops_patch = len(operations)
+        ops_replace = self._len_replace(playlist)
+
+        try:
+            if ops_replace < ops_patch:
+                self._replace_playlist(saved_playlist, new_tracks)
+            else:
+                self._patch_playlist(saved_playlist, operations)
+        except RuntimeError as e:
+            logger.error(f"Failed to save Spotify playlist: {e}")
+            # In the unlikely event that we used the replace strategy, and the
+            # first PUT went through but the following POSTs didn't, we have
+            # truncated the playlist. At this point, we still have the playlist
+            # data available, so we write it to an m3u file as a last resort
+            # effort for the user to recover from.
+            if ops_replace < ops_patch:
```

Comment:
> Why does it only get used in the replace strategy case?

because truncation can only happen when we use the replace strategy. i don't think its important to make a backup when patching the playlist. in the latter case, the user will see an error that their change failed, and the playlist is unmodified. but when using replace (and the very specific situation in the code comment happens; see last paragraph), the old playlist is clobbered, and we back it up in that case.

> And can't we do this for both playlists?

not sure what 'both'/other playlist you mean. the new state? this feature was only intended to avoid permanent data loss of previously existing data, hence i didn't implement saving the new state. (but see last big paragraph for more info)

>(why do we need length and artists? 

afaik, m3u-ext requires the length, and filling it in doesn't cost us anything. artist (and the whole extm3u block itself) were added to make the file human-parsable to an extent in this emergency situation.

>Should use Path. 

i think i am with `Extension.get_data_dir()`?

> Why binary mode? Should be ".m3u8"

i believe a remnant of a previous iteration. will change.

> and still needs a manual step to move the backup so Mopidy-M3U can find it. Could we consider something simpler? Maybe just dump (pickle?) both saved_playlist and playlist to disk? And then potentially adding a Spotify playlist restoration Command later?

i believe he likeliness of this code ever being called is extremely low. for it to trigger\*, we first have to make a successful PUT request, and then fail at a later POST request to the same endpoint, with the same authentication scope and the same parameters. and in normal use, replace mode will not be used at all, since no client supports batching multiple edit operations anyways (for a single edit, patch mode will always be used, since it's safer and requires the same amount of api requests). this is not something the user should have to deal with on a regular basis, so 'comfort features' are not included. consider this feature due diligence.

\* i'm actually creating the backup file not just when we truncate the playlist, but everytime something goes wrong with replace mode, since number of requests made is not available at this level. given that the code is pretty much never going to get called anyways, i didn't bother tightening down the check.

Reply:
> because truncation can only happen when we use the replace strategy.

OK, truncation is one class of issue but isn't data loss possible in either strategy? If a replacement is required and the remove operation succeeds but the corresponding add operation fails, won't that still leave you with a half modified playlist which is neither what we had or what we wanted? A backup of `playlist` in both cases would prevent that. Am I missing something?

> not sure what 'both'/other playlist you mean. the new state? this feature was only intended to avoid permanent data loss of previously existing data, hence i didn't implement saving the new state.

Your code here makes a backup of `playlist` (the requested new state), not `saved_playlist` (the old state). A request failure somewhere can leave us in a mess with our previous dat lost. So I was suggesting saving both playlists to give the best choice when it comes to recovery. 

> afaik, m3u-ext requires the length, 

No, there's no spec (yay!) and we (Mopidy-M3U) do not require it or even use it). A list of URIs is all that is required to be useful.

 > Should use Path.

Sorry, I meant [`pathlib.Path`](https://docs.python.org/3/library/pathlib.html) like the implementation in Mopidy-M3U. 

> since no client supports batching multiple edit operations anyways 

I think we've covered this before, few clients have any sort of playlist editing. We should support what is possible with the Mopidy API, not how we think people should/might/currently use it. Someone also uses it in a way you didn't think they would/should. 

> i believe he likeliness of this code ever being called is extremely low. for it to trigger*

Or we have a quota/rate limit issue half way through, or something else unlikely. But it if can happen it will happen. 

Maybe we should remove this code from the PR and address it later if anyone in the world ever sees (and reports) a problem.

Reply:
>truncation is one class of issue but isn't data loss possible in either strategy?

technically yes, but in my opinion only truncation is severe enough to care about. i have no problem with dropping the strategy test; will implement. 

>So I was suggesting saving both playlists to give the best choice when it comes to recovery.

will implement. 

> A list of URIs is all that is required to be useful.

useful to a computer; but i wanted it to be as useful as possible to humans as well. if we were to end up in this state, let's be as forthcoming to the user as possible.

>Sorry, I meant [`pathlib.Path`]

and [Extension.get_data_dir](https://docs.mopidy.com/en/latest/api/ext/#mopidy.ext.Extension.get_data_dir) is returning pathlib.path. we must really be talking about different things..?

>Maybe we should remove this code from the PR and address it later if anyone in the world ever sees (and reports) a problem.

strongly disagree. we should have some defence against truncation, since it's irreversible data loss. the trade-off is a few lines of imperfect code vs possible thousands of curated songs. if i were to trust mopidy-spotify with editing my years-old playlists, and it deleted most of their contents, and the maintainers just said "sorry, but thanks for the bugreport", I'd be majorly pissed. it can still be improved after committing.